Integrate cabal2bazel.

judah opened the pull request:
Modifies the previous version of Hazel to use `cabal2build`'s approach:
Have a Haskell file that that reifies the `.cabal` structure to `*.bzl`,
but otherwise write all logic directly as a Bazel macro.  (The previous
version had a more awkward round tripp between Haskell and Bazel.)

This PR is split into two commits:

1. A verbatim copy of certain files from `google/cabal2bazel`
2. Integration of those files into the existing `hazel` codebase.

There's more potential cleanup, but this seemed like a good milestone.

Currently supports:
- libraries and executables
- boot files
- preprocessors (happy/alex)
- data files
- `./configure` scripts
- `Paths_*` modules

In particular, builds a few nontrivial packages.  CI currently tests:
- `aeson`
- `language-c` (codegen from `happy`/`alex`)
- `lens` (dependencies use `Paths_` modules)
- `network` (`./configure` script)

Note: this PR manually checks in the data-files that Happy and Alex use
and which are generated automatically by their custom `Setup.hs` scripts.
Figuring out a better solution is TODO; however, those files' interfaces
have been *very* stable, and upcoming releases of those packages
will include them, so I think this is OK for now.

Some significant changes from `google/cabal2bazel`:

- The components of each package live in an autogenerated external repository
  (for example: `@haskell_{package}//...`), rather than
  `//third_party/haskell/{package}/...`.
- Default packages are taken from the `packages.bzl` file, rather than being
  hard-coded; they also need to be passed via the `builtin_dependencies`
  attribute.
- Package versions (needed for the Cabal macros file) are passed around
  manually by Bazel functions (since Hazel knows them up-front) rather than
  being passed through a separate `cabal_haskell_library` rule.
- Flag resolution code (`hazel_base_repository/Flatten.hs`) is implemented
  differently, to not require knowing the versions of packages
- Supports `Cabal>=2.0`.
- Don't set `CURRENT_PACKAGE_KEY` (I'm not aware of any packages using it)
- Use the Skylib `paths` module to clean up some code
- `hsc` files can be passed directly to build rules
- All source files need to be given a common `src_strip_prefix`; solve this
  by symlinking them all into a common location
- Locate data files correctly in the rule's external repository
- Work around `haskell_library` not supporting libraries without `srcs`

blackgnezdo commented:
I recommend a couple of things to make keeping track of things easy:
* use third_party convention
* preserve the original pristine commit (don't squash)

blackgnezdo commented:
Also your nice long PR comment should probably stay around as an .md file because it's a start of a design doc.

judah commented:
@blackgnezdo done; thanks for the suggestions!  I moved the cabal2bazel files into a third_party directory and also tried to keep their original structure for the time being.  Also added a Readme listing the differences.

blackgnezdo commented on third_party/cabal2bazel/Readme.md:
```diff
@@ -0,0 +1,33 @@
+#### Different semantics of `rules_haskell`
+- `hsc` files can be passed directly to build rules.
+- All source files need to be given a common `src_strip_prefix`; solve this
+  by symlinking them all into a common location.
```
I know you probably don't care at the moment. Is there a generic solution that bazel espouses for symlinking on Win?

judah replied:
I don't think so; there isn't really a generic "official" solution for symlinking on *nix systems either (see below).

blackgnezdo commented on third_party/cabal2bazel/bzl/cabal_package.bzl:
```diff
@@ -15,124 +15,58 @@
 _macros_suffix = "-macros"
 
+# Template files that we should install manually for Happy and Alex
+# TODO: figure out a better bootstrapping method
```
Maybe document it at least?

judah replied:
Done; added `templates/Readme.md`.

blackgnezdo commented on third_party/cabal2bazel/bzl/cabal_package.bzl:
```diff
@@ -15,124 +15,58 @@
 
+# Template files that we should install manually for Happy and Alex
+# TODO: figure out a better bootstrapping method
+_MANUAL_DATA_FILES = {
```
Does it make sense to host these lists in '@haskell_{happy,alex}//' instead of coupling things up tightly?

judah replied:
I moved the list into `templates/templates.bzl`.  Right now `@haskell_*` are auto-generated so the source of truth needs to live somewhere in this repo.

blackgnezdo commented on third_party/cabal2bazel/bzl/cabal_package.bzl:
```diff
@@ -189,7 +109,25 @@ _cabal_haskell_macros = rule(
+  ctx.actions.run(
+      outputs=[ctx.outputs.out],
+      inputs=[ctx.file.src],
+      executable="ln",
```
This suggests there's no generic abstraction in bazel to deal with symlinking. Oh well.

judah replied:
It seems like it at this point; see in particular https://github.com/bazelbuild/bazel/issues/4681
